doc/developer/design/20220718_envelopes.md

```diff
@@ -0,0 +1,208 @@
+in quotes because this does not correspond to a Kafka offset, but is instead a variety of metadata
+included inline with the data, which we _only_ know how to correctly interpret assuming the
+constraints described above are true.
+
+For deduplicating the initial snapshot in classic `DEBEZIUM` mode,
+we need to keep a set of seen keys, and drop any
+keys that are seen again. The reason for this is that there is no other known way of
+deduplicating messages from the initial snapshot, and Debezium is not known to emit any
+metadata that might be useful here (for example, there is no transaction ID or LSN
+corresponding to the records). It's okay not to durably store this data anywhere,
+since if we crash while ingesting the initial snapshot we can just restart from zero.
+
+The main wrinkle for classic `DEBEZIUM` mode is that not only do we need to persist
+the output data, but we also need to persist the in-memory deduplication state, since it's
+not possible to recover this without re-reading the upstream topic. This will require a sidecar
+"metadata" persist shard, which we must write to while emitting the data for every timestamp.
+On restart, we will re-ingest that state as of the restart timestamp
+
+The most difficult part of the project is supporting full consistency; i.e.,
+correctly interpreting the transaction metadata topic. Currently, that topic
+cannot be deduplicated without keeping O(k) in-memory state in the number of transations
+that have ever been processed in the lifetime of the upstream database; clearly,
+this cost will be unacceptable for many users. One possible approach is to
+change Debezium upstream to improve this situation; this is very feasible, but
+would mean we can't expect to support anyone but users of extremely recent Debezium versions.
+Another possibility is tolerating some amount of duplication; it's possible that we
+can still be correct while only using memory proportional to the _duplicated_ state,
+but this needs to be thought through in more detail.
+
+Another wrinkle with the transaction metadata topic is that we need custom logic for compacting it;
+briefly, we want to drop all data for transactions that have been fully processed,
+but this won't be possible if some of the transaction messages have been duplicated
+(and we only see the duplicate record after the transaction has been fully processed). Also, it has
+implications for creating new sources: if we've compacted away some past state, we won't be able to
+create new sources that are valid for those timestamps; thus we need some machinery for expressing
+that the `since` frontier of newly created sources depends on the frontier of their transaction
+topic. I spoke with Nikhil and Chae about this some and believe it's possible (though I'm
+still concerned about how it interacts with the deduplication story). At any rate it needs
+to be thought through further, which is why I've marked transaction topic support as a stretch goal.
```

Comment:
it being a stretch goal makes sense to me!  there's an implementation but there are still some known issues -- and it's behind `unsafe_mode` and undocumented so I don't think it's crucial to support right away